onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc

```diff
@@ -106,24 +108,15 @@ void XnnpackExecutionProvider::RegisterAllocator(AllocatorManager& allocator_man
     cpu_alloc = allocator_manager.GetAllocator(OrtMemTypeDefault, cpu_device);
 
     if (!cpu_alloc) {
-      // create our allocator
-      AllocatorCreationInfo allocator_info(
-          [](int) {
-            return std::make_unique<CPUAllocator>(OrtMemoryInfo(kXnnpackExecutionProvider,
-                                                                OrtAllocatorType::OrtDeviceAllocator));
-          });
-
-      cpu_alloc = CreateAllocator(allocator_info);
+      cpu_alloc = GetOrCreateAllocator().first;
       // enable sharing of our allocator
       allocator_manager.InsertAllocator(cpu_alloc);
     }
 
     InsertAllocator(cpu_alloc);
   }
 
-  // TODO: Create `struct xnn_allocator` that wraps cpu_allocator, and provide in the call to xnn_initialize so that
-  //       xnnpack is using the ORT allocator.
-  xnn_status st = xnn_initialize(nullptr);
+  const xnn_status st = xnn_initialize(GetOrCreateAllocator().second);
   if (st != xnn_status_success) {
```

Comment:
If we're sharing an allocator that already exists we want to wrap that. Currently this will always create a new allocator won't it?

Reply:
No, It won't.
1. `Allocator` is a static member in `GetOrCreateAllocator`, and `GetOrCreateAllocator().second` is a pointer of that xnn_allocator member.
2. `xnn_initialize` will save this pointer to an internal global variable by a thread-safe method, which guaranteed of `call_once`.

Reply:
If line 108 here returns an existing allocator we want to share that, so I think that's what we want to capture in the static (vs. always creating a new allocator by calling CreateAllocator in GetOrCreateAllocator). Otherwise you can end up with 2 CPU based allocators with different arenas and waste memory.

Reply:
You are right. It's a bug.

It doesn't matter to run `xnn_initialize` multiple times.  I am using a function to save the two static allocators.